### ISSUE TITLE
Support for afterFoundGet in HTTPDownloader

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4364]
|<img alt="zimmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| zimmer reported|
|-|-|
|Trac ID|trac#4364|
|Type|defect|
|Created|2010-03-11 02:19:41Z|

In twisted's current version HTTPDownloader [http://twistedmatrix.com/trac/browser/trunk/twisted/web/client.py#L353]
doesn't support afterFoundGet like HTTPClientFactory [http://twistedmatrix.com/trac/browser/trunk/twisted/web/client.py#L220] does. Here is a way to do (patch?) that.

Replace HTTPDownloader's _ _init_ _ with this
```
class HTTPDownloader(HTTPClientFactory):
    """Download to a file."""

    protocol = HTTPPageDownloader
    value = None

    def __init__(self, url, fileOrName,
                 method='GET', postdata=None, headers=None,
                 agent="Twisted client", supportPartial=0,
                 timeout=0, cookies=None, followRedirect=1,
                 redirectLimit=20, afterFoundGet = False):
        self.requestedPartial = 0
        if isinstance(fileOrName, types.StringTypes):
            self.fileName = fileOrName
            self.file = None
            if supportPartial and os.path.exists(self.fileName):
                fileLength = os.path.getsize(self.fileName)
                if fileLength:
                    self.requestedPartial = fileLength
                    if headers == None:
                        headers = {}
                    headers["range"] = "bytes=%d-" % fileLength
        else:
            self.file = fileOrName
        HTTPClientFactory.__init__(
            self, url, method=method, postdata=postdata, headers=headers,
            agent=agent, timeout=timeout, cookies=cookies,
            followRedirect=followRedirect, redirectLimit=redirectLimit,
            afterFoundGet = afterFoundGet)
```

This code (above) adds a new argument to the __init__ of HTTPDownloader (afterFoundGet) and then also passes it to the unbound __init__ of HTTPClientFactory.

Attachments:

* [after-found-get.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/984/984f1c55f733eb93231511e77e712caad0e15516/50c1f5cb21f3ece4490c9d421d492e6d2fb60a4a.diff) (2039 bytes) - added by magcius on 2011-02-13 14:33:02Z - Patch.

<details><summary>Searchable metadata</summary>

```
trac-id__4364 4364
type__defect defect
reporter__zimmer zimmer
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__HTTPClientFactory__HTTPDownloader__afterFoundGet__302_HTTP_Redirect HTTPClientFactory, HTTPDownloader, afterFoundGet, 302 HTTP Redirect
time__1268273981000000 1268273981000000
changetime__1297608595000000 1297608595000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks!  Patch looks good to me.  I'll apply it with just a couple minor tweaks:

  1. add a .feature news fragment describing the new parameter
  1. describe the behavior being tested more specifically in the test method docstring
  1. add an extra blank line after the new test method, so there are two between it and the next test method

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

(In [30712]) Add `afterFoundGet` parameter to `twisted.web.client.downloadPage`

Author: magcius
Reviewer: exarkun
Fixes: [#4364](https://github.com/twisted/twisted/issues/4364)

Add a parameter to `downloadPage` to allow controlling what type of request is issued after a `FOUND`
response is received in response to a request.  This `afterFoundGet` parameter works exactly the same
way the `getPage` parameter of the same name works.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
